src/scheduler/MpiWorld.cpp

```diff
@@ -906,64 +911,156 @@ void MpiWorld::gather(int sendRank,
+                     rankData.get(),
                      recvType,
-                     recvCount,
+                     recvCount * it.second.size(),
                      nullptr,
                      faabric::MPIMessage::GATHER);
+
+                // Copy each received chunk to its offset
+                for (int r = 0; r < it.second.size(); r++) {
+                    ::memcpy(recvBuffer + (it.second.at(r) * recvSize),
+                             rankData.get() + (r * recvSize),
+                             recvSize);
+                }
             }
         }
-    } else {
-        if (isInPlace) {
-            // A non-root rank running gather "in place" happens as part of an
-            // allgather operation. In this case, the send and receive buffer
-            // are the same, and the rank is eventually expecting a broadcast of
-            // the gather result into this buffer. This means that this buffer
-            // is big enough for the whole gather result, with this rank's data
-            // already in place. Therefore we need to send _only_ the part of
-            // the send buffer relating to this rank.
-            const uint8_t* sendChunk = sendBuffer + (sendRank * sendOffset);
-            send(sendRank,
-                 recvRank,
-                 sendChunk,
-                 sendType,
-                 sendCount,
-                 faabric::MPIMessage::GATHER);
-        } else {
-            // Normal sending
-            send(sendRank,
-                 recvRank,
-                 sendBuffer,
-                 sendType,
-                 sendCount,
-                 faabric::MPIMessage::GATHER);
+    } else if (isLocalLeader && !isLocalGather) {
+        // Scenario 2
+        auto rankData =
+          std::make_unique<uint8_t[]>(ranksForHost[thisHost].size() * sendSize);
+
+        // Gather all our local ranks data and send in a single remote message
+        for (int r = 0; r < ranksForHost[thisHost].size(); r++) {
+            if (ranksForHost[thisHost].at(r) == sendRank) {
+                // Receive from ourselves, just copy from/to the right offset
+                ::memcpy(rankData.get() + r * sendSize,
+                         sendBuffer + sendBufferOffset,
+                         sendSize);
+            } else {
+                // Receive from other local ranks
+                recv(ranksForHost[thisHost].at(r),
+                     sendRank,
+                     rankData.get() + r * sendSize,
+                     sendType,
+                     sendCount,
+                     nullptr,
+                     faabric::MPIMessage::GATHER);
+            }
         }
+
+        // Send the locally-gathered data to the receiver rank
+        send(sendRank,
+             recvRank,
+             rankData.get(),
+             sendType,
+             sendCount * ranksForHost[thisHost].size(),
+             faabric::MPIMessage::GATHER);
+
+    } else if (isLocalLeader && isLocalGather) {
```

Comment:
Scenario 3 and Scenario 5 are, after the re-factor, identical. However, they are, semantically, different situations.

I have decided to keep both as different branches for readability, and to help follow the code using the comment at the beginning.